### PR TITLE
oop_code

### PR DESCRIPTION
**Pull request by macoidk**
code is divided into classes and objects, all files are in the oop_code folder

### CODE REVIEW

File: k.makoyid/oop_code/Activation_Functions.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import numpy as np
+
+
+class ActivationFunctions:
+    @staticmethod
```
Comment (bulygin1985) [inline k.makoyid/oop_code/Activation_Functions.py]: Please add standard relu function. It is commonly used even in modern ML projects

File: k.makoyid/oop_code/Model_train.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        true_labels = np.argmax(Y, axis=0)
+        return np.mean(predictions == true_labels)
+
+    def train_model_with_lr_decay(self):
+        m = self.X_train.shape[1]
+        num_batches = m // self.batch_size
+
+        lr_schedule = [
+            (120, self.initial_lr),
+            (50, (self.initial_lr / 100) * 4),
```
Comment (bulygin1985) [inline k.makoyid/oop_code/Model_train.py]: Maybe, the second lr value has to be self.initial_lr / 10 (not self.initial_lr / 100) *4 ?

File: k.makoyid/oop_code/Model_train.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
+import numpy as np
+
+
+class ModelTrainer:
+    def __init__(self, model, X_train, Y_train, X_test, Y_test, initial_lr, batch_size, lambda_wd):
```
Comment (bulygin1985) [inline k.makoyid/oop_code/Model_train.py]: 1) add default values for  initial_lr, batch_size, lambda_wd (the optimal ones which you obtained).
2) add lr_schedule either as array [epoch_num_1, epoch_num_2, epoch_num_3 ,..] or as array of pairs [(epoc_num_1, factor_1), (epoch_num_2, factor_2), (epoch_3, factor_3), ...], where , factor__i is learning rate factor. Set some default value for lr_schedule

File: k.makoyid/oop_code/Network.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+class NeuralNetwork:
+    def __init__(self, input_size, hidden_size, num_classes):
+        self.W1 = np.random.randn(hidden_size, input_size) / np.sqrt(input_size)
+        self.b1 = np.zeros((hidden_size, 1))
+        self.W2 = np.random.randn(num_classes, hidden_size) / np.sqrt(hidden_size)
+        self.b2 = np.zeros((num_classes, 1))
+
+    def forward_propagation(self, X):
+        Z1 = np.dot(self.W1, X) + self.b1
+        A1 = ActivationFunctions.leaky_relu(Z1)
```
Comment (bulygin1985) [inline k.makoyid/oop_code/Network.py]: Set activation function as parameter in constructor (with default value  ActivationFunctions.leaky_relu)

File: k.makoyid/oop_code/Model_train.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        A2, _ = self.model.forward_propagation(X)
+        predictions = np.argmax(A2, axis=0)
+        true_labels = np.argmax(Y, axis=0)
+        return np.mean(predictions == true_labels)
+
+    def train_model_with_lr_decay(self):
+        m = self.X_train.shape[1]
+        num_batches = m // self.batch_size
+
+        lr_schedule = [
```
Comment (bulygin1985) [inline k.makoyid/oop_code/Model_train.py]: add add lr_schedule to constructor parameters (with default values)

**Review by bulygin1985 — CHANGES_REQUESTED** [review]

File: k.makoyid/oop_code/Activation_Functions.py (hunk 1 above)
Comment (constantinemakoid) [reply k.makoyid/oop_code/Activation_Functions.py]: according to the theory of the relu activation function, when the parameter alpha(that defines the slope of the function for negative inputs) = 0(now alpha = 0.01), this function will work as a standard relu

File: k.makoyid/oop_code/Model_train.py (hunk 2 above)
Comment (constantinemakoid) [reply k.makoyid/oop_code/Model_train.py]: during the training was found 3 lr`s which working better than other:
1) initial = 1e-3
2) then = 4e-4((50, (self.initial_lr / 100) * 4),)
3) 1e-7

**Comment by olnkravchenko** [comment]
@macoidk fix names of modules according to [PEP8 standards](https://peps.python.org/pep-0008/#package-and-module-names)

File: k.makoyid/oop_code/Activation_Functions.py (hunk 1 above)
Comment (olnkravchenko) [inline k.makoyid/oop_code/Activation_Functions.py]: There is no need in separate class, just delete it and leave functions as part of a module

File: k.makoyid/oop_code/Data_loader [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import numpy as np
+import tensorflow as tf
+from tensorflow.keras.datasets import cifar10
+
+
+class DataLoader:
+    @staticmethod
+    def load_cifar10_data():
```
Comment (olnkravchenko) [inline k.makoyid/oop_code/Data_loader]: What is the purpose of this class? Current implementation contains only one function which can be used without this class and it will be fine

Comment (olnkravchenko) [reply k.makoyid/oop_code/Data_loader]: Also, fix the file format

File: k.makoyid/oop_code/Model_train.py (hunk 3 above)
Comment (olnkravchenko) [inline k.makoyid/oop_code/Model_train.py]: Add type hint for `model` argument

File: k.makoyid/oop_code/Model_train.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+                    Y_batch = self.Y_train[:, start:end]
+
+                    A2, cache = self.model.forward_propagation(X_batch)
+                    gradients = self.model.backward_propagation(X_batch, Y_batch, cache, self.lambda_wd)
+                    self.model.update_parameters(gradients, lr)
+
+                current_epoch += 1
+                if current_epoch % 10 == 0:
+                    self.record_metrics()
+                    self.print_progress(current_epoch, lr)
```
Comment (olnkravchenko) [inline k.makoyid/oop_code/Model_train.py]: It should be done outside this function. Its main purpose is training a model. One option can be returning weights history as an output and then recalculate needed metrics using another class/function outside of this class

File: k.makoyid/oop_code/Model_train.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+        self.train_losses.append(train_loss)
+        self.test_losses.append(test_loss)
+        self.train_accuracies.append(train_accuracy)
+        self.test_accuracies.append(test_accuracy)
+
+    def print_progress(self, epoch, lr):
+        print(f"Epoch {epoch}, LR: {lr:.1e}, Train Loss: {self.train_losses[-1]:.4f}, "
+              f"Train Accuracy: {self.train_accuracies[-1]:.4f}, Test Loss: {self.test_losses[-1]:.4f}, "
+              f"Test Accuracy: {self.test_accuracies[-1]:.4f}")
```
Comment (olnkravchenko) [inline k.makoyid/oop_code/Model_train.py]: These functions violate Single Responsibility Principle. There should be only 1 purpose of a class - either train or record metrics & visualize them

File: k.makoyid/oop_code/Network.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        dW1 = 1 / m * np.dot(dZ1, X.T) + (lambda_wd / m) * self.W1
+        db1 = 1 / m * np.sum(dZ1, axis=1, keepdims=True)
+
+        return {"dW1": dW1, "db1": db1, "dW2": dW2, "db2": db2}
+
+    def update_parameters(self, gradients, learning_rate):
+        self.W1 -= learning_rate * gradients["dW1"]
+        self.b1 -= learning_rate * gradients["db1"]
+        self.W2 -= learning_rate * gradients["dW2"]
+        self.b2 -= learning_rate * gradients["db2"]
```
Comment (olnkravchenko) [inline k.makoyid/oop_code/Network.py]: Use for loop here

File: k.makoyid/oop_code/Visualizer.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import matplotlib.pyplot as plt
+
+
+class Visualizer:
+    @staticmethod
+    def plot_results(train_losses, train_accuracies, test_losses, test_accuracies):
+        plt.figure(figsize=(12, 8))
```
Comment (olnkravchenko) [inline k.makoyid/oop_code/Visualizer.py]: There is no need in class if it contains only static methods, especially this one

File: k.makoyid/oop_code/Visualizer.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+        plt.plot(test_losses, label='Test')
+        plt.xlabel('Iterations (x10)')
+        plt.ylabel('Loss')
+        plt.title('Loss vs Iterations')
+        plt.legend()
+
+        plt.subplot(222)
+        plt.plot(train_accuracies, label='Train')
+        plt.plot(test_accuracies, label='Test')
+        plt.xlabel('Iterations (x10)')
```
Comment (olnkravchenko) [inline k.makoyid/oop_code/Visualizer.py]: Will there always be x10 iterations? It's better to get this value as a function argument

File: k.makoyid/oop_code/main.py
Comment (olnkravchenko) [inline k.makoyid/oop_code/main.py:1]: It's better to use Jupyter Notebook in this case

File: k.makoyid/oop_code/Model_train.py (hunk 2 above)
Comment (bulygin1985) [reply k.makoyid/oop_code/Model_train.py]: So it is convenient to set lr_schedule as parameter in the class constructor (with default value)

**Review by bulygin1985 — APPROVED** [review]
